### PR TITLE
RUST-633 update driver to use tokio 1.0

### PR DESCRIPTION
**Pull request by saghm**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -60,18 +60,22 @@ webpki-roots = "0.18.0"
 version = "1.6.2"
 optional = true
 
+[dependencies.async-std-resolver]
```
Comment (saghm) [inline Cargo.toml:63]: This is trust-dns's implementation that uses async-std instead of tokio. Since there were slight changes to tokio's I/O traits that made it a little tricky to implement for async-std's types manually, I figured we might as well just switch to using the built-in support.

File: Cargo.toml [hunk 2]
```diff
@@ -21,12 +21,12 @@ exclude = [
...
 default = ["tokio-runtime"]
-tokio-runtime = ["tokio/dns", "tokio/macros", "tokio/rt-core", "tokio/tcp", "tokio/rt-threaded", "tokio/time", "reqwest", "serde_bytes"]
-async-std-runtime = ["async-std", "async-std/attributes"]
+tokio-runtime = ["tokio/macros", "tokio/net", "tokio/rt", "tokio/time", "reqwest", "serde_bytes"]
+async-std-runtime = ["async-std", "async-std/attributes", "async-std-resolver", "tokio-util/compat"]
 sync = ["async-std-runtime"]
 
 [dependencies]
-async-trait = "0.1.24"
+async-trait = "0.1.42"
```
Comment (saghm) [inline Cargo.toml:29]: In addition to updating tokio, I also updated all crates that use tokio under the hood, as well as async-trait

File: Cargo.toml [hunk 3]
```diff
@@ -83,13 +87,17 @@ version = "0.11.5"
...
-version = "~0.2.18"
+version = "1.0.1"
 features = ["io-util", "sync", "macros"]
 
 [dependencies.tokio-rustls]
-version = "0.13.0"
+version = "0.22.0"
 features = ["dangerous_configuration"]
 
+[dependencies.tokio-util]
```
Comment (saghm) [inline Cargo.toml:97]: This provides compatibility between futures versions of AsyncRead/AsyncWrite and the new tokio versions, which we need for our stream implementation

File: src/cmap/status.rs [hunk 4]
```diff
@@ -15,11 +13,7 @@ impl Default for PoolStatus {
 
 /// Create a channel for publishing and receiving updates to the pool's generation.
 pub(super) fn channel() -> (PoolGenerationPublisher, PoolGenerationSubscriber) {
-    let (sender, mut receiver) = tokio::sync::watch::channel(Default::default());
```
Comment (saghm) [inline src/cmap/status.rs:18]: Since the `watch` channel no longer is seeded with an initial message, this is no longer necessary

File: src/cmap/status.rs [hunk 5]
```diff
@@ -62,10 +56,11 @@ impl PoolGenerationSubscriber {
         timeout: std::time::Duration,
     ) -> Option<u32> {
         crate::RUNTIME
-            .timeout(timeout, self.receiver.recv())
+            .timeout(timeout, self.receiver.changed())
```
Comment (saghm) [inline src/cmap/status.rs:59]: The new API for the `watch` channel no longer yields a value on the receiving end directly. Instead, calling `changed` will await until a new value has been sent since the last time `changed` was called, and then `borrow` can be used to access the current value.

File: src/error.rs [hunk 6]
```diff
@@ -35,16 +35,12 @@ pub struct Error {
...
-        Error {
-            kind: e,
-            labels: Vec::new(),
-        }
-    }
-
     pub(crate) fn pool_cleared_error(address: &StreamAddress) -> Self {
         ErrorKind::ConnectionPoolClearedError {
-            message: format!("Conneciton pool for {} cleared during operation execution", address)
+            message: format!(
```
Comment (saghm) [inline src/error.rs:40]: There are a number of format changes that seems to have been made in files where other changes were made. I'm not sure why the format checking script didn't report anything when running rustfmt again seems to change the code.

File: src/error.rs [hunk 7]
```diff
@@ -68,20 +64,6 @@ impl Error {
         Error::authentication_error(mechanism_name, "invalid server response")
     }
 
-    /// Attempts to get the `std::io::Error` from this `Error`.
```
Comment (saghm) [inline src/error.rs:71]: This was only necessary for the custom trust-dns implementation, so it's not needed anymore

File: src/runtime/stream.rs [hunk 8]
```diff
@@ -74,9 +74,13 @@ impl AsyncTcpStream {
             timeout(connect_timeout, stream_future).await??
         };
 
-        stream.set_keepalive(Some(KEEPALIVE_TIME))?;
         stream.set_nodelay(true)?;
 
+        let socket = socket2::Socket::from(stream.into_std()?);
```
Comment (saghm) [inline src/runtime/stream.rs:79]: Tokio 1.0 currently doesn't expose any way to set keepalive on `TcpSocket`, but as a workaround, we can convert it to a `socket2` socket, set keepalive on that, and then convert it back.

File: src/runtime/stream.rs [hunk 9]
```diff
@@ -105,14 +109,6 @@ impl AsyncTcpStream {
         Ok(stream.into())
     }
 
-    pub(crate) async fn connect_socket_addr(
```
Comment (saghm) [inline src/runtime/stream.rs:108]: This was only used for the trust-dns custom implementation, so it isn't needed anymore

File: src/runtime/stream.rs [hunk 10]
```diff
@@ -165,12 +161,14 @@ impl AsyncRead for AsyncStream {
     fn poll_read(
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-        buf: &mut [u8],
+        mut buf: &mut [u8],
```
Comment (saghm) [inline src/runtime/stream.rs:164]: `tokio_util::io::poll_read_buf` requires an `&mut BufMut`, and `&mut [u8]` is a `BufMut`, so we need `&mut &mut [u8]`. Very confusing, I know...

File: src/runtime/stream.rs [hunk 11]
```diff
@@ -165,12 +161,14 @@ impl AsyncRead for AsyncStream {
...
     ) -> Poll<std::io::Result<usize>> {
         match self.deref_mut() {
             Self::Null => Poll::Ready(Ok(0)),
             Self::Tcp(ref mut inner) => AsyncRead::poll_read(Pin::new(inner), cx, buf),
-            Self::Tls(ref mut inner) => Pin::new(inner).poll_read(cx, buf),
+            Self::Tls(ref mut inner) => {
+                tokio_util::io::poll_read_buf(Pin::new(inner), cx, &mut buf)
```
Comment (saghm) [inline src/runtime/stream.rs:170]: Because tokio's AsyncRead trait was changed to use the new `ReadBuf` type instead of `[u8]`, we need to use an adapter to implement this trait.

File: src/runtime/stream.rs [hunk 12]
```diff
@@ -277,11 +275,21 @@ impl AsyncWrite for AsyncTcpStream {
...
-        AsyncRead::poll_read(self, cx, buf)
+        buf: &mut ReadBuf,
+    ) -> Poll<tokio::io::Result<()>> {
+        match self.deref_mut() {
+            #[cfg(feature = "tokio-runtime")]
+            Self::Tokio(ref mut inner) => Pin::new(inner).poll_read(cx, buf),
+
+            #[cfg(feature = "async-std-runtime")]
+            Self::AsyncStd(ref mut inner) => {
+                use tokio_util::compat::FuturesAsyncReadCompatExt;
```
Comment (saghm) [inline src/runtime/stream.rs:288]: Unfortunately `tokio_util` does not provide a one-off adapter function for the reverse direction, so to implement tokio's `AsyncRead` on an `async_std` stream, we need to wrap it in the `Compat` type that tokio_util provides.

File: src/sdam/description/server.rs [hunk 13]
```diff
@@ -79,7 +79,7 @@ pub(crate) struct ServerDescription {
 
 impl PartialEq for ServerDescription {
     fn eq(&self, other: &Self) -> bool {
-        if self.address == other.address && self.server_type == other.server_type {
+        if self.address != other.address || self.server_type != other.server_type {
```
Comment (saghm) [inline src/sdam/description/server.rs:82]: This is an unrelated bug that I found while debugging some test issues while this was in progress. Obviously, having the same address and server type does not mean that server descriptions are not equal.

File: src/sdam/description/topology/mod.rs [hunk 14]
```diff
@@ -85,7 +85,6 @@ impl PartialEq for TopologyDescription {
         self.compatibility_error == other.compatibility_error
             && self.servers == other.servers
             && self.topology_type == other.topology_type
-            && self.cluster_time == other.cluster_time
```
Comment (saghm) [inline src/sdam/description/topology/mod.rs:88]: This is another unrelated bug; we should not be considering topology descriptions to be different if only the cluster time is different, or else we'll be notifying the server selection operations that they should wake up to try to select a server when nothing that actually matters to them has changed.

File: src/sdam/message_manager.rs [hunk 15]
```diff
@@ -1,51 +1,47 @@
-use std::{sync::Arc, time::Duration};
+use std::time::Duration;
 
-use tokio::sync::watch::{self, Receiver, Sender};
+use tokio::sync::broadcast::{self, Receiver, Sender};
```
Comment (saghm) [inline src/sdam/message_manager.rs:3]: While I was debugging some issues with this due to some quirks with the `watch` channel, Patrick suggested that the `broadcast` channel might make more sense here, since we fundamentally want a multi-producer, multi-consumer channel. Switching to that makes the code much cleaner, and since we didn't use the internals of this type from outside the module, it also didn't require changing any outside code.

**Review by patrickfreed — COMMENTED** [review]
The code changes all look good to me, but the test failures do look like they need to be addressed.

1. It looks like `tokio-util` doesn't compile on our MSRV, so we'll need to update it in 2.0 to whatever tokio requires. I looked briefly through their documentation and didn't see anything mentioning what that actually is though. I'm sure its out there somewhere.
2. There are a few server selection failures ([1](https://evergreen.mongodb.com/task_log_raw/mongo_rust_driver_tests__async_runtime~async_std_auth_and_tls~noauth_and_notls_os~ubuntu_16.04_test_latest_replica_set_patch_3639656719a77436125c8320b29cd907226a641a_600f2dbde3c331565dfcf0b2_21_01_25_20_44_46/0?type=T#L1386), [2](https://evergreen.mongodb.com/task_log_raw/mongo_rust_driver_tests__async_runtime~async_std_auth_and_tls~noauth_and_notls_os~windows_64_vs2017_small_test_3.6_replica_set_patch_3639656719a77436125c8320b29cd907226a641a_600f2dbde3c331565dfcf0b2_21_01_25_20_44_46/0?type=T#L1677)) that claim a `ReplicaSetWithPrimary` topology isn't suitable for a Primary read preference, which seems like a bug. Given that the topology should be selectable, it seems like an issue of the server selection tasks/threads not getting woken up when the topology changes properly. I wonder if there's a race between server selection failing and changes to the topology being subscribed to or something. I know you mentioned this in standup but I'm not sure if the latest commit was meant to address it or not.

On a more meta-note, I think we'll want to branch off a 1.x-master/main style branch at some point, along with a new evergreen project. We could probably just re-use the old change streams evergreen for that purpose.

Oh, and could you add RUST-633 to the PR title?

**Comment by saghm** [comment]
Yep, I had mentioned that tokio-util didn't compile on 1.43 on Slack the other day. Updating our minimum in a major version bump shouldn't be an issue.

I agree that those failures are something we'll want to fix before a stable 2.0 release, but the last commit I made seems to have eliminated them on tokio, only leaving them on async-std. Obviously we'll want to run more tests to verify this, but assuming that's the case, I don't think it necessarily needs to be a blocker for the 2.0 alpha; the main reason we're putting out an alpha is to allow users to use tokio 1.0 immediately, so users who are using async-std shouldn't have any strong need to use the alpha when the only changes compared to the next feature release of 1.x that we're doing are to update tokio to 1.0.

**Comment by patrickfreed** [comment]
Ok that sounds good. Could you file a bug report so we remember to fix it before 2.0? We'll also need to fix it in 1.x as per the maintenance guidelines. Could you also file ticket(s) for the bugs fixed in this PR so we remember to backport them to 1.x as well?

**Review by patrickfreed — APPROVED** [review]

File: src/error.rs (hunk 6 above)
Comment (isabelatkinson) [reply src/error.rs:40]: Tiny nit, but while we're here, can you fix the typo in this message (connection spelled incorrectly)?

**Review by isabelatkinson — APPROVED** [review]
lgtm mod one typo

**Comment by Darksonn** [comment]
> I looked briefly through their documentation and didn't see anything mentioning what that actually is though. I'm sure its out there somewhere.

The minimum supported version is 1.45.
